src/core/batch.rkt

```diff
@@ -189,48 +197,51 @@
-                   (add-enode (eggref iff) type))
-             (list 'if
-                   (add-enode (eggref cond) 'bool)
-                   (add-enode (eggref ift) type)
-                   (add-enode (eggref iff) type)))]
-        [(list (? impl-exists? impl) ids ...)
-         (define args
-           (for/list ([id (in-list ids)]
-                      [type (in-list (impl-info impl 'itype))])
-             (add-enode (eggref id) type)))
-         (cons impl args)]
-        [(list (? operator-exists? op) ids ...)
-         (define args
-           (for/list ([id (in-list ids)]
-                      [type (in-list (operator-info op 'itype))])
-             (add-enode (eggref id) type)))
-         (cons op args)]))
-    (batch-push! out node*))
-
-  (define (finalize-batch roots)
-    (set! input-batch (mutable-batch->batch out (list->vector roots)))
-    (for/list ([root (in-list roots)])
-      (batchref input-batch root)))
-
-  (values add-enode finalize-batch))
+  (define (add-enode enode type)
+    (define idx
+      (let loop ([enode enode]
+                 [type type])
+        (define enode*
+          (match enode
+            [(? number?) (if (representation? type) (literal enode (representation-name type)) enode)]
+            [(? symbol?)
+             (if (hash-has-key? rename-dict enode) (car (hash-ref rename-dict enode)) enode)]
+            [(list '$approx spec (app eggref impl))
+             (define spec* (vector-ref id->spec spec))
+             (unless spec*
+               (error 'regraph-extract-variants "no initial approx node in eclass"))
+             (define spec-type (if (representation? type) (representation-type type) type))
+             (define final-spec (egg-parsed->expr spec* rename-dict spec-type))
+             (define final-spec-idx (mutable-batch-munge! out final-spec))
+             (approx final-spec-idx (loop impl type))]
+            [(list 'if (app eggref cond) (app eggref ift) (app eggref iff))
+             (if (representation? type)
+                 (list 'if (loop cond (get-representation 'bool)) (loop ift type) (loop iff type))
+                 (list 'if (loop cond 'bool) (loop ift type) (loop iff type)))]
+            [(list (? impl-exists? impl) (app eggref args) ...)
+             (define args*
+               (for/list ([arg (in-list args)]
+                          [type (in-list (impl-info impl 'itype))])
+                 (loop arg type)))
+             (cons impl args*)]
+            [(list (? operator-exists? op) (app eggref args) ...)
+             (define args*
+               (for/list ([arg (in-list args)]
+                          [type (in-list (operator-info op 'itype))])
+                 (loop arg type)))
+             (cons op args*)]))
+        (batch-push! out enode*)))
+    (batchref input-batch idx))
+
+  ; same as add-enode but works with index as an input instead of enode
+  (define (add-id id type)
+    (add-enode (eggref id) type))
```

Comment:
`add-id` is used in `regraph-extract-best` because it extracts based on the index in egg's nodes

Reply:
While `add-enode` is used in `regraph-extract-variants` since it extracts using `enode`